packages/core/versioning/core-versioned-http-server-internal/src/internal_versioned_route.ts

```diff
@@ -0,0 +1,164 @@
+    public readonly options: VersionedRouteConfig<Method>,
+    // TODO: Make "true" dev-only
+    private readonly validateResponses: boolean = true
+  ) {}
+
+  /** This is where we must implement the versioned spec once it is available */
+  private requestHandler = async (
+    ctx: RequestHandlerContextBase,
+    req: KibanaRequest,
+    res: KibanaResponseFactory
+  ) => {
+    const version = req.headers[VERSION_HEADER] as undefined | Version;
+    if (!version) {
+      return res.custom({
+        statusCode: 406,
+        body: `Version expected at [${this.method}] [${this.path}]. Please specify a version in the ${VERSION_HEADER} header.`,
+      });
+    }
+
+    const handler = this.handlers.get(version);
+    if (!handler) {
+      return res.custom({
+        statusCode: 406,
+        body: `No ${version} available for ${this.method} ${this.path}. Available versions are: ${[
+          this.handlers.keys(),
+        ].join(',')}`,
+      });
+    }
+
+    const coreKibanaRequest = req as CoreKibanaRequest;
+    if (handler.options.validate && handler.options.validate.request) {
+      try {
+        const { body, params, query } = validate(
+          req,
+          handler.options.validate.request,
+          handler.options.version
+        );
+        coreKibanaRequest.body = body;
+        coreKibanaRequest.params = params;
+        coreKibanaRequest.query = query;
+      } catch (e) {
+        return res.custom({
+          statusCode: 400,
+          body: e.message,
+        });
+      }
+    } else {
+      // Preserve behavior of not passing through unvalidated data
+      coreKibanaRequest.body = {};
+      coreKibanaRequest.params = {};
+      coreKibanaRequest.query = {};
+    }
+
+    const result = await handler.fn(ctx, req, res);
+
+    if (this.validateResponses && handler.options.validate && handler.options.validate.response) {
+      const { response } = handler.options.validate;
+      try {
+        validate(
+          req,
+          { body: response.body, unsafe: { body: response.unsafe } },
+          handler.options.version
+        );
+      } catch (e) {
+        return res.custom({
+          statusCode: 500,
+          body: `Failed output validation: ${e.message}`,
+        });
+      }
+    }
+
+    return result;
+  };
+
+  public addVersion(options: Options, handler: RequestHandler<any, any, any, any>): VersionedRoute {
+    if (this.handlers.has(options.version)) {
+      throw new Error(
+        `Version ${
+          options.version
+        } handler has already been registered for the route "${this.method.toLowerCase()} ${
+          this.path
+        }"`
+      );
+    }
+
+    this.handlers.set(options.version, { fn: handler, options });
+    if (this.handlers.size === 1) {
+      this.router[this.method](
+        {
+          path: this.path,
+          validate: passThroughValidation,
+          options: this.options,
+        },
+        this.requestHandler
+      );
+    }
```

Comment:
Given that we're returning a specific http code/message for invalid/unknown version, I think we want to register the route as soon as the versioned route is created (so in the constructor, basically), rather than waiting for the first version to be registered to do so.

My concern would be to avoid a difference of behavior when:
- requesting `v2` to a route without any version registered
and
- `requesting `v2` to a route with only `v1` registered